src/integration-test/java/com/commercetools/sync/integration/services/impl/UnresolvedReferencesServiceImplIT.java

```diff
@@ -0,0 +1,155 @@
+import static io.sphere.sdk.utils.SphereInternalUtils.asSet;
+import static java.util.Collections.singleton;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class UnresolvedReferencesServiceImplIT {
+
+    private UnresolvedReferencesService unresolvedReferencesService;
+
+    private List<String> errorCallBackMessages;
+    private List<String> warningCallBackMessages;
+    private List<Throwable> errorCallBackExceptions;
+
+    private static final String CUSTOM_OBJECT_CONTAINER_KEY =
+        "commercetools-sync-java.UnresolvedReferencesService.productDrafts";
+
+
+    @BeforeEach
+    void setup() {
+        deleteWaitingToBeResolvedCustomObjects(CTP_TARGET_CLIENT);
+    }
+
+    @BeforeEach
+    void setupTest() {
+        errorCallBackMessages = new ArrayList<>();
+        errorCallBackExceptions = new ArrayList<>();
+        warningCallBackMessages = new ArrayList<>();
+
+        final ProductSyncOptions productSyncOptions = ProductSyncOptionsBuilder.of(CTP_TARGET_CLIENT)
+            .errorCallback(
+                (errorMessage, exception) -> {
+                    errorCallBackMessages
+                            .add(errorMessage);
+                    errorCallBackExceptions
+                            .add(exception);
+                })
+            .warningCallback(warningMessage ->
+                    warningCallBackMessages
+                            .add(warningMessage))
+            .build();
+
+        unresolvedReferencesService = new UnresolvedReferencesServiceImpl(productSyncOptions);
+    }
+
+    @Test
+    void save_WithoutException_createsNewCustomObject() {
+        // preparation
+        final ProductDraft productDraft =
+            SphereJsonUtils.readObjectFromResource(PRODUCT_KEY_1_RESOURCE_PATH, ProductDraft.class);
+
+        final WaitingToBeResolved productDraftWithUnresolvedRefs =
+                new WaitingToBeResolved(productDraft, asSet("foo", "bar"));
+
+        // test
+        final Optional<WaitingToBeResolved> result = unresolvedReferencesService
+            .save(productDraftWithUnresolvedRefs)
+            .toCompletableFuture()
+            .join();
+
+        // assertions
+        assertThat(result).hasValueSatisfying(waitingToBeResolved ->
+            assertThat(waitingToBeResolved.getProductDraft()).isEqualTo(productDraft));
+
+        final CustomObjectByKeyGet<WaitingToBeResolved> customObjectByKeyGet = CustomObjectByKeyGet
+            .of(CUSTOM_OBJECT_CONTAINER_KEY, productDraft.getKey(), WaitingToBeResolved.class);
+        final CustomObject<WaitingToBeResolved> createdCustomObject = CTP_TARGET_CLIENT
+            .execute(customObjectByKeyGet)
+            .toCompletableFuture()
+            .join();
+
+        assertThat(createdCustomObject.getValue()).isEqualTo(productDraftWithUnresolvedRefs);
+    }
+
+    @Test
+    void fetchCustomObject_shouldReturnCorrectCustomObject() {
```

Comment:
Can't you test fetching also in the previous test? You're saving the custom object and then you need to fetch it to assert, so why not to test it there also?

Reply:
Also deleting could be done in one test 😄 since it's an integration test, you should rather less tests if possible to save times and resources.

Reply:
e6adfd4